src/types.ts

```diff
@@ -104,17 +104,37 @@ export type Blockchain =
       ens?: NamingServiceSource;
       zns?: NamingServiceSource;
       cns?: NamingServiceSource;
-      web3Provider?: Web3Provider;
+      web3Provider?: Provider;
+      provider?: Provider;
     };
 
 export type API = {
   url: string;
 };
 
-export interface Web3Provider {
+export interface Provider {
   sendAsync: (method: string, params: any) => Promise<any>;
 }
 
+export interface JsonRpcPayload {
+  jsonrpc: string;
+  method: string;
+  params: any[];
+  id?: string | number;
+}
+
+export interface JsonRpcResponse {
+  jsonrpc: string;
+  id: number;
+  result?: any;
+  error?: string;
+}
+
+export interface AbstractProvider {
+  sendAsync?: (payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void) => void;
+  send?: (payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void) => void;
```

Comment:
Woun't there be one required property here or maybe there should be at least one? You can describe this situation with typescript like:

```
type ProviderMethod = (payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void) => void
type AbstratProvider = {sendAsync: ProviderMethod} | {send: ProviderMethod}
```

Reply:
It seems that the newer providers have been migrated from using send method to sendAsync method.
That means depending on which provider the client is using they might either not have sendAsync at all and have send method or vise-versa 

If I do it in a way above the user will be getting typescript error saying that the provider I am sending is not the same as being expected since sendAsync is not implemented for older providers like  web3-providers-ws or web3-providers-http  and it won't compile for them without hard type casting.

I think the method we used now is the safest since we do manually check for both methods anyway and in case none found we throw the IncorectProvider Error with the message that neither send nor sendAsync is implemented in the provider.

Reply:
To be more clear this is what I am getting on our side when defining AbstractProvider via | sign, I would assume the client-side would get a similar error when sending a provider to us

<img width="1012" alt="Screen Shot 2020-06-11 at 4 04 12 PM" src="https://user-images.githubusercontent.com/16522024/84447615-50dda500-abfd-11ea-82f3-571e6bca26fa.png">

Or maybe I am not doing something right.

Reply:
@bogdan Some of the deprecated web3 providers (like `web3-providers-http` that @JohnnyJumper mentioned, not `web3-providers-http` though) implement only `send` method. See https://github.com/ethereum/web3.js/blob/1.x/packages/web3-core-requestmanager/src/index.js#L162